File: backend/ai/schemas.py
from pydantic import BaseModel, field_validator, model_validator
from typing import Literal


class ScenarioCase(BaseModel):
    probability: float
    target: str
    thesis: str

    @field_validator("probability")
    @classmethod
    def probability_in_range(cls, v: float) -> float:
        if not 0.0 <= v <= 1.0:
            raise ValueError("probability must be between 0 and 1")
        return v


class AIAnalysisResult(BaseModel):
    bull_case: ScenarioCase
    base_case: ScenarioCase
    bear_case: ScenarioCase
    risk_factors: list[str]
    max_drawdown_estimate: str
    position_size_pct: float
    confidence: float
    timeframe: str

    @field_validator("confidence")
    @classmethod
    def confidence_in_range(cls, v: float) -> float:
        if not 0.0 <= v <= 1.0:
            raise ValueError("confidence must be between 0 and 1")
        return v

    @model_validator(mode="after")
    def probabilities_sum_to_one(self) -> "AIAnalysisResult":
        total = (
            self.bull_case.probability
            + self.base_case.probability
            + self.bear_case.probability
        )
        if abs(total - 1.0) > 0.05:
            raise ValueError(
                f"bull + base + bear probabilities must sum to ~1.0 "
                f"(within 0.05 tolerance), got {total:.4f}"
            )
        return self


class ScreeningResult(BaseModel):
    ticker: str
    action: Literal["BUY", "HOLD", "SELL", "WATCH"]
    conviction: float
    one_liner: str

    @field_validator("conviction")
    @classmethod
    def conviction_in_range(cls, v: float) -> float:
        if not 0.0 <= v <= 1.0:
            raise ValueError("conviction must be between 0 and 1")
        return v


class MarketOutlook(BaseModel):
    regime_assessment: str
    sector_rotation: list[str]
    risk_level: str
    key_themes: list[str]
    outlook_text: str

File: backend/src/ai/schemas.ts
/**
 * Zod schemas for AI output validation (replaces Pydantic models).
 */

import { z } from 'zod';

export const ScenarioCaseSchema = z.object({
  probability: z.number().min(0).max(1),
  target: z.string(),
  thesis: z.string(),
});

export type ScenarioCase = z.infer<typeof ScenarioCaseSchema>;

export const AIAnalysisResultSchema = z
  .object({
    bull_case: ScenarioCaseSchema,
    base_case: ScenarioCaseSchema,
    bear_case: ScenarioCaseSchema,
    risk_factors: z.array(z.string()),
    max_drawdown_estimate: z.string(),
    position_size_pct: z.number(),
    confidence: z.number().min(0).max(1),
    timeframe: z.string(),
  })
  .refine(
    (data) => {
      const total =
        data.bull_case.probability +
        data.base_case.probability +
        data.bear_case.probability;
      return Math.abs(total - 1.0) <= 0.05;
    },
    {
      message:
        'bull + base + bear probabilities must sum to ~1.0 (within 0.05 tolerance)',
    },
  );

export type AIAnalysisResult = z.infer<typeof AIAnalysisResultSchema>;

export const ScreeningResultSchema = z.object({
  ticker: z.string(),
  action: z.enum(['BUY', 'HOLD', 'SELL', 'WATCH']),
  conviction: z.number().min(0).max(1),
  one_liner: z.string(),
});

export type ScreeningResult = z.infer<typeof ScreeningResultSchema>;

export const MarketOutlookSchema = z.object({
  regime_assessment: z.string(),
  sector_rotation: z.array(z.string()),
  risk_level: z.string(),
  key_themes: z.array(z.string()),
  outlook_text: z.string(),
});

export type MarketOutlook = z.infer<typeof MarketOutlookSchema>;

export const PortfolioReviewSchema = z.object({
  investment_stage: z.string(),
  portfolio_grade: z.enum(['A', 'B', 'C', 'D', 'F']),
  overall_assessment: z.string(),
  strengths: z.array(z.string()),
  weaknesses: z.array(z.string()),
  recommendations: z.array(z.string()),
  risk_alerts: z.array(z.string()),
});

export type PortfolioReview = z.infer<typeof PortfolioReviewSchema>;
